### DISCUSSION TITLE
[Question] Deno and Node future.

### DISCUSSION THREAD

**Discussion by diego9497** [prisma/prisma#2415] [Q&A] [answered]
What do you think about Deno?
Will prisma be compatible with deno "easily"?

I don't know too much about Prisma itself and less of Deno, but I loved Prisma and I want to use it in the future. And I want to know if add deno support will be to much complicated.

Thanks to all the team.

**Comment by ryands17** [comment, marked as answer]
Hey @diego9497 :wave: 
A great question indeed! 
Deno looks great and has some exciting features, but there are a lot of things to consider yet like how would the exosystem look like and would it be able to interop with the existing `npm` packages or would we have to create a separate one for Deno.

It would be best to look into Deno after it has achieved some point of stability and usage so it wouldn't be on our list right now, but you can surely open a feature request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that we could look into it in the future.

**Reply by kotx** [reply]
> would it be able to interop with the existing `npm` packages

not much experience with deno but https://esm.sh/ might be helpful?

**Reply by andykais** [reply]
it would if prisma were simple es6 modules, but prisma has native modules that need to build a bunch of things at install time (think rust dependencies). I believe theyre currently using node-gyp and deno doesnt have a built-in method for building native dependencies out of the box.

**Reply by kotx** [reply]
Perhaps using prebuilt binaries would be a temporary workaround?

**Reply by janpio** [reply]
Prisma is downloading prebuilt engine files (binaries), no `node-gyp` around here.

**Reply by andykais** [reply]
>. Prisma is downloading prebuilt engine files

thats great! There would still need to be installer logic for deno though, essentially at some point you would need to fetch the binaries and then link them into a deno script. Something like:
```ts
let suffix = "so";
if (Deno.build.os == "windows") suffix = "dll"
const binary_name = `./${os}-${version}.${suffix}`
const binary_response = await fetch(`https://hosted-prisma-binaries.com/${binary_path}`)
const binary_file = `./${binary_name}`
await Deno.writeFile(binary_file, await blob.arrayBuffer())
const prisma_interface = { // this would need to manually define all the native interfaces
  connect_to_database: { parameters: ["string"], result: "i32" }
}
const dylib = Deno.dlopen(binary_file, prisma_interface)

class PrismaClient {
  constructor() {
    dylib.connect_to_database('./path.db')
   }
}

export {
  PrismaClient
}
```

**Comment by Gabriel-Louis** [comment]
The Question asked is really great, no doubt deno is great, it take years for a framework, runtime to solidify it ground. Node has a large community, when would deno catch up with node.. Obviously it would take years!

**Comment by andykais** [comment]
Minimally, we would want `@prisma/client` supported in deno. The way I see it, there are three things to tackle:
1. use [denoify](https://github.com/garronej/denoify) to transpile typescript files into deno compatible files
2. find out how to convert rust/native dependencies (node-gyp built deps) with either wasm imports, or bind to the new [FFI](https://deno.com/blog/v1.13#experimental-ffi-replaces-native-plugin-system) plugin system.
3. create a CI pipeline to publish to deno.land/x. This stage is straightforward.

For (1) to work, prisma has to only use nodejs apis that are supported by deno's compatibility layer https://deno.land/std@0.104.0/node#supported-builtins (denoify should handle mapping this for us)
For (2), Ill admit Im unfamiliar with a lot of this stuff. I know prisma uses rust modules internally, which are fairly well supported by wasm https://developer.mozilla.org/en-US/docs/WebAssembly/Rust_to_wasm.  Alternatively, we could create platform specific dynamic libraries that deno loads. This is more ugly of course, because deno would have to fetch & load these at runtime

**Comment by ludralph** [comment]
Hi @diego9497,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
